plugins/reindex/src/main/java/org/elasticsearch/plugin/reindex/AbstractAsyncBulkByScrollAction.java

```diff
@@ -0,0 +1,347 @@
+    private final ThreadPool threadPool;
+    private final SearchRequest firstSearchRequest;
+    private final ActionListener<Response> listener;
+
+    public AbstractAsyncBulkByScrollAction(ESLogger logger, Client client, ThreadPool threadPool, Request mainRequest,
+            SearchRequest firstSearchRequest, ActionListener<Response> listener) {
+        this.logger = logger;
+        this.client = client;
+        this.threadPool = threadPool;
+        this.mainRequest = mainRequest;
+        this.firstSearchRequest = firstSearchRequest;
+        this.listener = listener;
+    }
+
+    protected abstract BulkRequest buildBulk(Iterable<SearchHit> docs);
+
+    protected abstract Response buildResponse(long took);
+
+    public void start() {
+        initialSearch();
+    }
+
+    /**
+     * Count of documents updated.
+     */
+    public long updated() {
+        return updated.get();
+    }
+
+    /**
+     * Count of documents created.
+     */
+    public long created() {
+        return created.get();
+    }
+
+    /**
+     * Count of successful delete operations.
+     */
+    public long deleted() {
+        return deleted.get();
+    }
+
+    /**
+     * The number of scan responses this request has processed.
+     */
+    public int batches() {
+        return batches.get();
+    }
+
+    public long versionConflicts() {
+        return versionConflicts.get();
+    }
+
+    public long successfullyProcessed() {
+        return updated.get() + created.get() + deleted.get();
+    }
+
+    public List<Failure> failures() {
+        return unmodifiableList(failures);
+    }
+
+    private void initialSearch() {
+        try {
+            startTime.set(System.nanoTime());
+            if (logger.isDebugEnabled()) {
+                logger.debug("executing initial scroll against {}{}",
+                        firstSearchRequest.indices() == null || firstSearchRequest.indices().length == 0 ? "all indices"
+                                : firstSearchRequest.indices(),
+                        firstSearchRequest.types() == null || firstSearchRequest.types().length == 0 ? ""
+                                : firstSearchRequest.types());
+            }
+            client.search(firstSearchRequest, new ActionListener<SearchResponse>() {
+                @Override
+                public void onResponse(SearchResponse response) {
+                    logger.debug("[{}] documents match query", response.getHits().getTotalHits());
+                    onScrollResponse(response);
+                }
+
+                @Override
+                public void onFailure(Throwable e) {
+                    logger.error("failed while executing the initial scroll request", e);
+                    listener.onFailure(e);
+                }
+            });
+        } catch (Throwable t) {
+            finishHim(t);
+        }
+    }
+
+    void onScrollResponse(SearchResponse searchResponse) {
+        threadPool.generic().execute(() -> {
```

Comment:
My plan was to use BulkProcessor but then that got lost in the shuffle. I believe handling rejections should still be a requirement for merging to master so its worth a NOCOMMIT.